user_input.pl

```diff
@@ -20,16 +24,94 @@
 % Get the difficulty level as input from the user
 pick_difficulty_level(Level):-
     writeln("Please select a difficulty level:"),
-    writeln("1- Beginner"),
-    writeln("2- Intermediate"),
-    writeln("3- Expert"),
+    writeln("1: Beginner"),
+    writeln("2: Intermediate"),
+    writeln("3: Expert"),
     repeat,
-    read(UserInput),
+    get_single_char(UserInputCode),
+    number_codes(UserInput, [UserInputCode]),
     (
-        integer(UserInput),
         (UserInput = 1, Level = easy), !;
         (UserInput = 2, Level = intermediate), !;
-        (UserInput = 3, Level = expert), !;
+        (UserInput = 3, Level = expert), !,
+        retractall(difficulty_level(_)),
+        assert(difficulty_level(Level)), !;
         writeln("Invalid input. Please enter a valid difficulty level."),
         fail
-    ).
\ No newline at end of file
+    ).
+
+% Get the location of the ball the player chooses to move next. The ball chosen
+% will be the one that pushes the other balls in the convoy (if there are any).
+pick_ball_to_move(BoardState, Player, Row, Column):-
+    board_size(BoardSize),
+    writeln("Pick a ball to move. The ball will push other balls in the direction you choose."),
+    writeln("Row number:"),
+    repeat,
+    (
+        pick_row_number(BoardSize, Row),
+        writeln("Column letter:"),
+        pick_col_number(BoardSize, Column),
+        slot_by_index(BoardState, Row, Column, Player), !;
+        writeln("Invalid input. Please pick a ball as your color."),
+        fail
+    ).
+
+% Get a valid row number from the user. It must be within the
+% bounds of the board.
+pick_row_number(BoardSize, Row):-
+    repeat,
+    (
+        read(UserInput),
+        between(1, BoardSize, UserInput),
+        Row = UserInput, !;
+        writeln("Invalid input. Please enter a valid row number."),
+        fail
+    ).
+
+% Get a valid column number from the user. It must be within the
+% bounds of the board.
+pick_col_number(BoardSize, Col):-
+    repeat,
+    (
+        read(UserInput),
+        UpperLimit is 65 + BoardSize,
+        between(65, UpperLimit, UserInput),
+        Col is UserInput - 64, !;
```

Comment:
We should either use our `col_to_num` method in conversions like this or delete it because it is so simple yet adds more lines of code. I think I'm in favor of deleting it.